omod/src/main/java/org/openmrs/web/controller/ForgotPasswordFormController.java

```diff
@@ -143,26 +152,46 @@ protected ModelAndView onSubmit(HttpServletRequest request, HttpServletResponse
 				try {
 					Context.addProxyPrivilege(PrivilegeConstants.GET_USERS);
 					user = Context.getUserService().getUserByUsername(username);
-				}
-				finally {
+				} finally {
 					Context.removeProxyPrivilege(PrivilegeConstants.GET_USERS);
 				}
 				
+				String secretQuestion = null;
+				if (user != null) {
+					secretQuestion = Context.getUserService().getSecretQuestion(user);
+				}
+				
 				// check the secret question again in case the user got here "illegally"
-				if (user == null /*|| StringUtils.isEmpty(user.getSecretQuestion())*/) {
+				if (user == null) {
+					httpSession.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, "auth.answer.invalid");
+					httpSession.setAttribute(WebConstants.OPENMRS_MSG_ATTR, "auth.question.fill");
+					request.setAttribute("secretQuestion", getRandomFakeSecretQuestion(username));
+				} else if (secretQuestion == null || secretQuestion.equals("")) {
 					httpSession.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, "auth.question.empty");
-				} else if (/*user.getSecretQuestion() != null &&*/ Context.getUserService().isSecretAnswer(user, secretAnswer)) {
+				} else if (secretQuestion != null && Context.getUserService().isSecretAnswer(user, secretAnswer)) {
 					
 					StringBuilder randomPassword = new StringBuilder();
-					for (int i = 0; i < 8; i++) {
-						randomPassword.append(String.valueOf((Math.random() * (127 - 48) + 48)));
+					ArrayList<Character> buildRandomPassword = new ArrayList<Character>();
+					
+					//Ensure atleast 1 uppercase, 1 lowercase as well as 1 digit in password (Otherwise, exception is thrown)
+					buildRandomPassword.add((char) (65 + (int) Math.abs(Math.random()) % 26));
+					buildRandomPassword.add((char) (97 + (int) Math.abs(Math.random()) % 26));
+					buildRandomPassword.add((char) (48 + (int) Math.abs(Math.random()) % 10));
+					//Build rest of the password
+					for (int i = 3; i < 8; i++) {
+						buildRandomPassword.add((char) (Math.random() * (127 - 48) + 48));
+					}
+					//Shuffle it
+					Collections.shuffle(buildRandomPassword);
+					//Finally, append it to the string
+					for (char characterInPassword : buildRandomPassword) {
+						randomPassword.append(characterInPassword);
```

Comment:
Before you ask me to change the password code, I tried doing it with RandomStringUtils but I kept getting an error. I think I'll stick to this. Seems to work out fine. I have tested it